### PR TITLE
Configfile functions and GlobalConfig.json example

### LINKED ISSUE #6: Create Set-Config
Sets a json config file in a standard location.  This will need to be determined when building get/set config functions.

### PR DESCRIPTION
**Pull request by thoen008**
Resolves #6 #5 #16 Created 2 functions Get-Config and Set-Config to solve 6 and 5. Created an example globalconfig.JSON file for solving 16.

I have tested each of the functions as best as I can putting various values in the required fields for Set-Config. I validated the -verbose logging is working.

### CODE REVIEW

File: UMN-AutoPackager/public/Set-Config.ps1
Comment (gjslavik) [inline UMN-AutoPackager/public/Set-Config.ps1:34]: Maybe use the new Get-Config function here to make things more dynamic and future-proof?

File: GlobalConfig.json
Comment (jbolduan) [inline GlobalConfig.json:8]: I think we should add a second item to each location for location type so maybe something like this:
```suggestion
  "RecipeLocations": [
    {
        "locationType": "directory",
        "locationUri": "C:\\Temp"
    },
    {
        "locationType": "directory",
        "locationUri": "\\\\files.somewhere\\"
    }
```

This will make adding additional locations in the future easier (GitHub)

File: GlobalConfig.json
Comment (jbolduan) [inline GlobalConfig.json:1]: We should add paths for where software should be downloaded to locally and where applications files should be stored for SCCM.

I think it would also make sense to have the SCCM be an array much like the suggestion I have for recipeLocations so we can have multiple targets.

File: UMN-AutoPackager/public/Get-GlobalConfig.ps1
Comment (jbolduan) [inline UMN-AutoPackager/public/Get-GlobalConfig.ps1:1]: This should be made Get-GlobalConfig or Get-ModuleConfig (maybe Get-UMNAutopackagerConfig to be unique?)

File: UMN-AutoPackager/public/Get-GlobalConfig.ps1
Comment (jbolduan) [inline UMN-AutoPackager/public/Get-GlobalConfig.ps1:25]: I would add in the -InputObject switch for the $config variable.  Not required but makes things cleaner.

File: UMN-AutoPackager/public/Get-GlobalConfig.ps1
Comment (jbolduan) [inline UMN-AutoPackager/public/Get-GlobalConfig.ps1:23]: I would add in the -Path for the $path variable to make this cleaner.

File: UMN-AutoPackager/public/Set-GlobalConfig.ps1
Comment (jbolduan) [inline UMN-AutoPackager/public/Set-GlobalConfig.ps1:1]: See Get-Config, this name should be changed to match the decision made on that name.

**Review by jbolduan — CHANGES_REQUESTED** [review]
Looks good overall, I've suggested a few changes.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (gjslavik) [inline UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1:77]: For future revisions, we may want to add functionality to ensure paths have / and \ escaped either as validated input or when writing values to the json file.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (gjslavik) [inline UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1:80]: As Jeff mentioned on one of my scripts, there are a lot of Write-Verbose lines. Even when running in non-Verbose mode, these still process and output to void. To save on processing power, they should be used where necessary, rather then where they are nice to have. I did the same thing on mine and pared them down a bit.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (gjslavik) [inline UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1:153]: I think this last line was from your testing and you might not have intended to include it.

File: UMN-AutoPackager/public/Set-Config.ps1
Comment (thoen008) [reply UMN-AutoPackager/public/Set-Config.ps1]: Added the function get-UMNGlobalconfig.

File: GlobalConfig.json
Comment (thoen008) [reply GlobalConfig.json]: Updated the JSON with new arrays for ConfigMgr and RecipeLocations.

File: GlobalConfig.json
Comment (thoen008) [reply GlobalConfig.json:1]: Added additional information for these to the ConfigMgr array.

File: UMN-AutoPackager/public/Get-GlobalConfig.ps1
Comment (thoen008) [reply UMN-AutoPackager/public/Get-GlobalConfig.ps1]: renamed the functions to Get-UMNGlobalConfig and Set-UMNGlobalConfig

File: UMN-AutoPackager/public/Get-GlobalConfig.ps1
Comment (thoen008) [reply UMN-AutoPackager/public/Get-GlobalConfig.ps1]: Reviewed the code and added any missing switches. I think I got them all.

File: UMN-AutoPackager/public/Get-GlobalConfig.ps1
Comment (thoen008) [reply UMN-AutoPackager/public/Get-GlobalConfig.ps1]: Added switches to any I saw that were missing.

File: UMN-AutoPackager/public/Set-GlobalConfig.ps1
Comment (thoen008) [reply UMN-AutoPackager/public/Set-GlobalConfig.ps1]: Done.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (thoen008) [reply UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1]: Commented out some of the write-verbose to reduce the number.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (thoen008) [reply UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1]: Yeah. I was committing changes while still working out the functionality. So you probably saw one of those iterations. I have removed them in the last commit.

File: UMN-AutoPackager/public/Set-Config.ps1
Comment (gjslavik) [reply UMN-AutoPackager/public/Set-Config.ps1]: Great! Resolving.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (gjslavik) [reply UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1]: Thanks! Resolving.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (gjslavik) [reply UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1]: That's what I figured. Resolving.

File: UMN-AutoPackager/public/Get-UMNGlobalConfig.ps1
Comment (jbolduan) [inline UMN-AutoPackager/public/Get-UMNGlobalConfig.ps1:12]: We should standardize around help being outside the function right above the delcaration.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (jbolduan) [inline UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1:32]: We should standardize around help being outside the function right above the delcaration.

**Review by jbolduan — CHANGES_REQUESTED** [review]
Looks good, only change I would request (open to discussion) is moving help to right above the function declaration.  It makes functions more readable.

File: UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1
Comment (gjslavik) [reply UMN-AutoPackager/public/Set-UMNGlobalConfig.ps1]: Resolving for now. If it needs to be an Enhancement later, we can open an issue for it then. This doesn't need to hold up approval. Resolving.

**Comment by gjslavik** [comment]
Updates look good. I am OK with comments inside the Function braces to align with the PowerShell Style Guide.
Approving.

**Review by gjslavik — APPROVED** [review]
See previous comments regarding Approval.

**Review by jbolduan — APPROVED** [review]
We agreed that comment help should be inside the functions based on community coding standards.

**Pull request merged**
